Tutorial USP

btrb opened the pull request:
This PR is for including the material developed for the workshop in USP on Devito's tutorial/examples.

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
```diff
@@ -0,0 +1,363 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In Devito, an `Operator` carries out three fundamental tasks: generation of low-level code, JIT compilation, and execution. Overall, the lowering process performed by an `Operator` – from high-level equations to dynamically compiled and executable code – consists of multiple compiler passes, summarized in the Figure bellow. The following series of notebooks will expose those compilation stag ...
```
there is no figure below

have you checked that this premise is not redundant in the other notebooks ?

btrb replied:
please see comment bellow

btrb replied:
added link as pdf!

btrb replied:
i noticed the pdf was not being displayed.
i found [this](https://stackoverflow.com/questions/19470099/view-pdf-image-in-an-ipython-notebook) solution, but i'm not sure if i'm overkilling it.
would you have any advice on it?

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
```diff
@@ -0,0 +1,363 @@
+  {
+   "attachments": {
+    "compiler_stages.png": {
+     "image/png": "iVBORw0KGgoAAAANSUhEUgAAAzQAAAKaCAYAAADoExxdAAAABHNCSVQICAgIfAhkiAAAABl0RVh0U29mdHdhcmUAZ25vbWUtc2NyZWVuc2hvdO8Dvz4AACAASURBVHic7N15WFXl/v7x90YEAQERVCAcUTE1iRTFzBwiM4fjkGOZcw6Zx1NmpXkcMsscTnrM4XiyPP1skBwqc9bEKRWcEsxUVEQQRURRcWLD+v3Bl60IKCqyQe/XdXG1XftZz3OvtcT2Zz9rMBmGYSAiIiIiIlIE2Vg7gIiIiIiIyP1SQSMiIiIiIkWWChoRERERESmyVNCIiIiIiEiRpYJGRERERESKLBU0IiIiIiJSZNlaO4CISKbz588TERFBbGws ...
```
typo: functionf

btrb replied:
i have hosted the image so i dont have to export it as a string like that. please see new commit to see if that solved

btrb replied:
added image as pdf, please see comment above

btrb commented:
ok, just created a new `00_index.ipynb` explaining what's going on in a MVP fashion. 

about the file `lowering`, it is still incomplete. i will add more about `subs` and `lowering` yet. 
please see new commit to see if that's clarifying

georgebisbas commented on examples/compiler/00_index.ipynb:
```diff
@@ -0,0 +1,53 @@
+    "\n",
+    "The following series of notebooks will expose key aspects of those compilation stages in a hands-on fashion. We will be exploring the heart of the `Operator` class, which can be found at `devito/operator.py`.\n",
+    "\n",
+    "The examples and exercices shown here are mostly taken either from tests found in the folder `devito/tests`, or from the documentation itself.\n",
```
exercises

FabioLuporini replied:
I think we can actualy safely drop this sentence, or at least be more conservative ("Many of the examples shown in these notebooks are extracted from...") but I'd not say all of them or even mostly.

btrb replied:
thank you @georgebisbas , fixed.

ok @FabioLuporini fixed.

georgebisbas commented on examples/compiler/01_lowering.ipynb:
spatial

btrb replied:
thank you

georgebisbas commented on examples/compiler/01_lowering.ipynb:
Sparse

FabioLuporini replied:
I'd entirely drop "Indexifying functions allows us to write finite differences stencils from it."

and then:

"can be indexified in two equivalent ways, either via the `.indexify()` method [and then you show the notebook cell] or via explicit indexing, as typically done in user code [and then you add a new cell showing that ``f[time, x] -> f[time, x]``]"

btrb replied:
thank you george

@FabioLuporini fixed partially. I didn't get what you mean by "or via explicit indexing, as typically done in user code [and then you add a new cell showing that f[time, x] -> f[time, x]]"

georgebisbas commented on examples/compiler/01_lowering.ipynb:
through

btrb replied:
thank you

georgebisbas commented:
> Nice work, trivial comments.
> Can that be used or extended to cover #520 ?

ok, sorry, I guess you have already that into your mind! ;-) Just saw your #520 talk...

FabioLuporini commented on examples/compiler/00_index.ipynb:
```diff
@@ -0,0 +1,53 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# The Devito's Operator \n",
```
Drop 's I think. Or maybe better "The Devito compiler"

btrb replied:
fixed

FabioLuporini commented on examples/compiler/00_index.ipynb:
```diff
@@ -0,0 +1,53 @@
+    "\n",
+    "In Devito, an `Operator` carries out three fundamental tasks: generation of low-level code, JIT compilation, and execution. Overall, the lowering process performed by an `Operator` – from high-level equations to dynamically compiled and executable code – consists of multiple compiler passes, summarized in the Figure bellow. \n",
+    "\n",
+    "![Compiler Stages](https://i.ibb.co/dKjsZ72/compiler-stages.png)\n",
```
I'd still prefer it to be a pdf, and I'd make it smaller

btrb replied:
done

FabioLuporini commented on examples/compiler/00_index.ipynb:
```diff
@@ -0,0 +1,53 @@
+    "\n",
+    "![Compiler Stages](https://i.ibb.co/dKjsZ72/compiler-stages.png)\n",
+    "\n",
+    "The following series of notebooks will expose key aspects of those compilation stages in a hands-on fashion. We will be exploring the heart of the `Operator` class, which can be found at `devito/operator.py`.\n",
```
> will expose key aspects of those compilation stages in a hands-on fashion

I'd rephrase it as

"will describe these compilation passes in a hands-on fashion"

Then:

"We will explore the heart .."  ("will be exploring" here sounds odd)

Then another thing:

"of the `Operator` class, ..."

two comments here:

1) I'd write "of the `Operator` class, which orchestrates code generation and compilation."
2) I'd add a hover link to devito/operator.py -- I think the syntax is \[Operator class\](https://github.com/opesci/devito/blob/master/devito/operator.py)"

btrb replied:
fixed

FabioLuporini commented on examples/compiler/00_index.ipynb:
```diff
@@ -0,0 +1,53 @@
+    "## Outline\n",
+    "- 01) Lowering\n",
+    "    - Indexification\n",
+    "    - Substitution\n",
```
Is this a different notebook than indexification?

again, here you could add links to the actual notebooks through the usual syntax \[item\](link)

btrb replied:
no, i thought of it as in the same notebook

btrb replied:
about these links, i'm not sure how to proceed regarded that they are not in mater yet, so the shortest way i found was to add the link for my this current branch. like:

`01) [Lowering](https://github.com/opesci/devito/blob/master/examples/compiler/01_iet.ipynb)`

FabioLuporini replied:
no worries, you can add them in a later PR otherwise

FabioLuporini replied:
see my comment below for more details -- but basically we can infer what the link names will be

btrb replied:
ok

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
"... Functions into Indexeds, that is actual array accesses. An Indexed ..."

btrb replied:
fixed

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
no need for save=1 here right?

btrb replied:
no, i'm just avoiding having to explain why `u.data` would have a buffer at this point

FabioLuporini replied:
but save=1 will still raise a question I think. Anyway, fine!

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
Dimensions

FabioLuporini replied:
read -> interpreted?

btrb replied:
thank you

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
I'd drop this and rewrite as "Clearly, we now have that [and show cell]"

btrb replied:
fixed

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
Yet, -> However,
those -> these

btrb replied:
thank you

btrb commented:
i'm still facing errors with the CI for the case `gcc8`. has someone experienced this ?

FabioLuporini commented on examples/compiler/02_data_regions.ipynb:
```diff
@@ -11,7 +11,7 @@
    "metadata": {},
    "source": [
-    "In this tutorial we will learn about data regions and how these impact the Operator construction. We will use the simple time marching example shown in the [01_iet](https://github.com/opesci/devito/blob/master/examples/compiler/01_iet.ipynb) tutorial. "
+    "In this tutorial we will learn about data regions and how these impact the Operator construction. We will use the simple time marching example."
```
why this change?

btrb replied:
because this filename will change.
not only that, but probably this notebook will come first than the notebook on the IET (see the index in [00_index.ipynb](https://github.com/opesci/devito/blob/tutorial_usp-2/examples/compiler/00_index.ipynb)

FabioLuporini replied:
then we need to move it in here, don't we?

also, I think we can infer what the filename will be -- it's pretty much the same string that is in current master, except for the actual filename that changes. So I'd use it and as soon as the branch lands, it should work; then we go check, and if we broke something, we can immediately file and merge a hotfix PR to tweaks the broken link (but again, this might not even be an issue)

btrb replied:
ok

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
maybe we should call it ``time``, not ``t``. Since here you're using ``save=...``, we get ``time``, not ``t``, in the dimensions list.

btrb replied:
fixed

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
again, probably should be ``time, x = ...``

btrb replied:
ok

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
The explicit indexing mode is when you write
```
a = u[time, x]
```
and you get exactly (representation shown below)
```
[In 0] a
[Out 0] u[time, x]
```
it totally equivalent to indexify. Another example, you can write
```
b = u[time+1, x-2]
```
and you get
```
[In 1] b
[Out 1] u[time+1, x-2]
```
and clearly you can still do things like ``b.function`` to get ``u(time, x)``

FabioLuporini replied:
I do know you know about it, but I just wrote this short summary to clarify the terminology.

Also I spotted a TYPO above: bellow -> below

btrb replied:
thank you! added this to the notebook.

FabioLuporini commented:
Maybe it is convenient to put ``substitutions`` and ``domain-alignment`` in a different notebook, so that we can merge this PR ASAP. As there will be more to be added under ``lowering`` (see the new method in master's operator.py, called "_add_implicit",  to deal with subdomains...). Also, I **think** the data regions notebook could/should be moved before the ``domain-alignment`` one. What do you think?

btrb commented:
> Maybe it is convenient to put `substitutions` and `domain-alignment` in a different notebook, so that we can merge this PR ASAP. As there will be more to be added under `lowering` (see the new method in master's operator.py, called "_add_implicit", to deal with subdomains...). Also, I **think** the data regions notebook could/should be moved before the `domain-alignment` one. What do you think?

roger that. fixed this for a quick PR.
i'm still stuck with the `gcc8` issue. it seems that some cells of other notebooks are failing in this particular case

i completely agree with moving data regions to before of domain alignment.

FabioLuporini commented on examples/compiler/00_index.ipynb:
```diff
@@ -0,0 +1,51 @@
+    "Many of the examples and exercises shown here are extracted from test cases found in the folder `devito/tests`, or from the documentation itself.\n",
+    "\n",
+    "## Outline\n",
+    "- 01) [Indexification](https://github.com/opesci/devito/blob/master/examples/compiler/01_indexification.ipynb)\n",
```
the notebook names do not match. Maybe call this `01_lowering_indexification`, and change the filename from 01_lowering into 01_lowering_indexification ?

btrb replied:
typo! thank you, fixed

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
`u[t + 1, x − 2]`would  -> leave space before "would"

btrb replied:
fixed

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
time,x -> time, x

btrb replied:
ok, fixed

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
time,x -> time, x

btrb replied:
fixed

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
in fact, this could and should be ``a.function is b.function``

in short:
* ``is`` will check for **identity**, i.e., that they are the same object. Indeed, if you look at ``id(a.function)`` and ``id(b.function)`` you'll see that they are exactly the same
* ``==`` will instead check for **equality** -- two objects may be equal (based on the ``__eq__`` method) but they could be different objects

and here the function is exactly the same object, so we can use a stronger (and also more efficient) comparison (namely, ``is``)

btrb replied:
oh! thank you for that
fixed

FabioLuporini commented on examples/compiler/01_lowering.ipynb:
`grid` or `Grid` ? 
there's no variable called `grid` in this notebook, so I suppose you want `Grid`

btrb replied:
i was trying to refer to the parameter grid of the object `TimeFunction`, but i did as you suggested.

FabioLuporini commented on examples/compiler/02_data_regions.ipynb:
```diff
@@ -11,15 +11,13 @@
    "metadata": {},
    "source": [
-    "In this tutorial we will learn about data regions and how these impact the Operator construction. We will use the simple time marching example shown in the [01_iet](https://github.com/opesci/devito/blob/master/examples/compiler/01_iet.ipynb) tutorial. "
+    "In this tutorial we will learn about data regions and how these impact the `Operator` construction. We will use the simple time marching example."
```
reminder:
>  We will use the simple time marching example."
... which one ? :)
probably you need to lift it from 03_iet...

btrb replied:
imho, i think that linking the something in the intro of the second notebook for references that live in the third notebook is somewhat confusing. (please note that the notebooks have shifted in order. `iet` is no longer the first notebook)

 also, the a time marching example is described in cell `[03]` of this very notebook

FabioLuporini replied:
ahhh I had missed this...

> also, the a time marching example is described in cell [03] of this very notebook

this sentence was confusing:

>  We will use the simple time marching example

you should probably rewrite it as

>  We will use a simple time-marching example

(or Operator instead of example doesn't matter)

FabioLuporini replied:
I mean, the article ('the' instead of 'a') was confusing

btrb replied:
of course, agreed

FabioLuporini commented:
> point 2: i'm still not sure if you can visualize the pdf of 00_index.ipynb that i've uploaded from there. may you confirm it please?

I cannot !

> point 1: i'm still stuck with the gcc8 case in CI. it seems that some cells of the notebook 01 are failing. all errors are:
Unexpected output fields from running code: {'text/latex'} (like for cell [1])
have you seen this before?

try using `print(X)` instead of just `X` ?

btrb commented:
> > point 2: i'm still not sure if you can visualize the pdf of 00_index.ipynb that i've uploaded from there. may you confirm it please?
> 
> I cannot !

fixed this with
```
from IPython.display import IFrame
```

> 
> > point 1: i'm still stuck with the gcc8 case in CI. it seems that some cells of the notebook 01 are failing. all errors are:
> > Unexpected output fields from running code: {'text/latex'} (like for cell [1])
> > have you seen this before?
> 
> try using `print(X)` instead of just `X` ?

thank you, trying that

btrb commented:
it seems the solution using `Iframe` wont pass through CI. 
should I use .png instead, or maybe other solution for displaying .pdf?

FabioLuporini commented:
@btrb 

> it seems the solution using Iframe wont pass through CI.
should I use .png instead, or maybe other solution for displaying .pdf?

My suggestion is, instead of working around, just try to understand what the issue is. Have you looked at the CI logs? the error seems to be a stupid one -- just the Python ``repr`` of an object changing (as expected) at each run. 
  
'<IPython.lib...7f50a8919320>' == '<IPython.lib....7f09f844ee10>'
  - <IPython.lib.display.IFrame at 0x7f50a8919320>
  ?                                    - ^ ^ ---
  + <IPython.lib.display.IFrame at 0x7f09f844ee10>
  ?                                     ^^ ^^^^

So this would be easily fixable -- either assigning the output to a dummy variable or using the usual ``#NBVAL_IGNORE_OUTPUT`` at the top of the cell.]

However, why do we not just insert the image via Markdown? As for example shown [here](https://stackoverflow.com/questions/32370281/how-to-embed-image-or-picture-in-jupyter-notebook-either-from-a-local-machine-o)

this way, the cell will be a markdown one, and so won't cause any harm to CI (CI, via the ``nbval`` package, only runs the Python cells, not the Markdown ones)

FabioLuporini commented on examples/compiler/00_index.ipynb:
```diff
@@ -0,0 +1,85 @@
+   "source": [
+    "# The Devito compiler\n",
+    "\n",
+    "The following series of notebooks will describe these compilation passes in a hands-on fashion. We will explore the heart of the `Operator` class, which orchestrates code generation and compilation. For more details, see the [Operator class](https://github.com/opesci/devito/blob/master/devito/operator.py).\n",
```
"these compilation passes" -- which ones? the sentence is a bit odd...

btrb replied:
oh yes. i shouldn't have moved those paragraphs orders in the first place.

FabioLuporini commented on examples/compiler/00_index.ipynb:
```diff
@@ -0,0 +1,85 @@
+   "source": [
+    "Many of the examples and exercises shown here are extracted from test cases found in the folder `devito/tests`, or from the documentation itself.\n",
+    "\n",
+    "## Outline\n",
```
Could be a good time to write out the entire index, clearly with links to the available notebooks only.
The index, also, should/could have a tree structure. For example, to start with (we could revisit this in future PRs) we could have


* Preliminaries
    * [Data regions](https://github.com/opesci/devito/blob/master/examples/compiler/02_data_regions.ipynb)
    * Data dependence analysis
* Lowering
    * [Indexification](https://github.com/opesci/devito/blob/master/examples/compiler/01_lowering.ipynb)
    * Substitutions
    * Domain alignment
    * Eq -> LoweredEq
* Clustering
* Symbolic optimization via the Devito Symbolic Engine (DSE)
* Tree-fication
    * Clusters -> ScheduleTree
    * ScheduleTree -> Iteration/Expression Tree (IET)
    * IET examples
        * [Part A: top-down inspection](https://github.com/opesci/devito/blob/master/examples/compiler/03_iet-A.ipynb)
        * [Part B: bottom-up construction](https://github.com/opesci/devito/blob/master/examples/compiler/04_iet-B.ipynb)
* IET analysis
* Loop optimization via the Devito Loop Engine (DLE)
    * Classic transformations (SIMD, blocking, ...)
    * OpenMP parallelism
    * MPI parallelism
* Finalization     

which is rendered via
```
* Preliminaries
    * [Data regions](https://github.com/opesci/devito/blob/master/examples/compiler/02_data_regions.ipynb)
    * Data dependence analysis
* Lowering
    * [Indexification](https://github.com/opesci/devito/blob/master/examples/compiler/01_lowering.ipynb)
    * Substitutions
    * Domain alignment
    * Eq -> LoweredEq
* Clustering
* Symbolic optimization via the Devito Symbolic Engine (DSE)
* Tree-fication
    * Clusters -> ScheduleTree
    * ScheduleTree -> Iteration/Expression Tree (IET)
    * IET examples
        * [Part A: top-down inspection](https://github.com/opesci/devito/blob/master/examples/compiler/03_iet-A.ipynb)
        * [Part B: bottom-up construction](https://github.com/opesci/devito/blob/master/examples/compiler/04_iet-B.ipynb)
* IET analysis
* Loop optimization via the Devito Loop Engine (DLE)
    * Classic transformations (SIMD, blocking, ...)
    * OpenMP parallelism
    * MPI parallelism
* Finalization
```

yes, as you can see here I'm wondering whether the "data regions" notebook actually deserves a separate (preliminary) chapter...

Clearly this is just a suggestion ! What do you think?

btrb replied:
yes! that's what i've thought 
ofc this is way ahead lol ! 
thank you for making it more clear

FabioLuporini replied:
glad to help!

FabioLuporini commented on examples/compiler/02_data_regions.ipynb:
```diff
@@ -33,7 +31,7 @@
    "metadata": {},
    "source": [
-    "At this point, we have a time-varying 3x3 grid filled with `1's`. Below, we can see the `domain` data values:"
+    "At this point, we have a time-varying 3x3 grid filled with _1's_. Below, we can see the _domain_ data values:"
```
I like this change. In practice, now domain/halo/padding are written in italics, while the actual devito objects (such as TimeFunction, Operator, etc) within ` . Correct?

btrb replied:
exactly

FabioLuporini replied:
nice

btrb commented:
> @btrb
> 
> > it seems the solution using Iframe wont pass through CI.
> > should I use .png instead, or maybe other solution for displaying .pdf?
> 
> My suggestion is, instead of working around, just try to understand what the issue is. Have you looked at the CI logs? the error seems to be a stupid one -- just the Python `repr` of an object changing (as expected) at each run.
> 
> '<IPython.lib...7f50a8919320>' == '<IPython.lib....7f09f844ee10>'
> 
> * <IPython.lib.display.IFrame at 0x7f50a8919320>
>   ?                                    - ^ ^ ---
> 
> * <IPython.lib.display.IFrame at 0x7f09f844ee10>
>   ?                                     ^^ ^^^^
> 
> So this would be easily fixable -- either assigning the output to a dummy variable or using the usual `#NBVAL_IGNORE_OUTPUT` at the top of the cell.]

going for the 2nd

> 
> However, why do we not just insert the image via Markdown? As for example shown [here](https://stackoverflow.com/questions/32370281/how-to-embed-image-or-picture-in-jupyter-notebook-either-from-a-local-machine-o)

just tried this, but it looks that this example doesn't work for `.pdf`s

> 
> this way, the cell will be a markdown one, and so won't cause any harm to CI (CI, via the `nbval` package, only runs the Python cells, not the Markdown ones)

FabioLuporini commented:
@btrb I think this is fantastic work! Many thanks for all this, really appreciated. 

There only are two small items left, and then we merge this:

* I tried running the notebook, and now to see the PDF we have to execute the cell...which is a bit counterintuitive. I think I prefer to have a JPG/PNG/whatever embedded directly in markdown. Even though it's not PDF, it is still better than having to execute a cell to see a static image
* The notebook enumeration is now wrong, since we have flipped data_regions with indexification.

btrb commented:
> @btrb I think this is fantastic work! Many thanks for all this, really appreciated.
> 
> There only are two small items left, and then we merge this:
> 
> * I tried running the notebook, and now to see the PDF we have to execute the cell...which is a bit counterintuitive. I think I prefer to have a JPG/PNG/whatever embedded directly in markdown. Even though it's not PDF, it is still better than having to execute a cell to see a static image

ok, fixed

> * The notebook enumeration is now wrong, since we have flipped data_regions with indexification.

you're right. in fact, i renamed `lowering` to `indexification` in the new commit

FabioLuporini commented:
@btrb there are few more things that need to be addressed. I'm writing them down here, but only one is strictly needed for this PR. I can take care of the others as I assume you're getting tired of this :D

* the only thing we have to do is to resize the image in 00_index and make it way smaller, as it's currently a punch in the eye (also, at such a big size, the resolution really sucks)

The other non-compulsory things to do are:

* The beginning of 03_iet-A is now a bit off. This sentence
```
This is the first part of a tutorial describing the Iteration/Expression Tree (IET), an intermediate representation based on Abstract Syntax Trees (AST) used in the Devito compiler. In this part we will show how to access and navigate the IET rooted in an Operator.
```
Given the previous (existing and non-existing) notebooks, we should rewrite it simply as
```
In this tutorial we will show how to access and navigate the Iteration/Expression Tree (IET) rooted in an `Operator`.
```

* This sentence
```
The reader of this tutorial is expected to be familiar with the basic Devito API (i.e., Grid, Function/TimeFunction, Operator, ...). Otherwise, the CFD tutorials in examples/cfd are a better place to start.
```
should/could be moved to 00_index (and replace "of this tutorial" with "of this series of tutorials")

* Similarly, the beginning of 04_iet-B should be updated. From
```
This is the second part of a tutorial describing the Iteration/Expression Tree (IET), an intermediate representation based on Abstract Syntax Trees (AST) used in the Devito compiler. In this part we will learn how to build, compose, and transform IETs.
```
to something simpler
```
In this tutorial we will learn how to build, compose, and transform Iteration/Expression Trees (IETs).
```

btrb commented:
> @btrb there are few more things that need to be addressed. I'm writing them down here, but only one is strictly needed for this PR. I can take care of the others as I assume you're getting tired of this :D
> 
oh no, please don't assume that. i'm actually looking forward to push it further as soon as this PR is merged.

> * the only thing we have to do _in this PR_ is to resize the image in 00_index and make it (much) smaller, as it's currently a punch in the eye (also, at such a big size, the resolution really sucks)
> 

ok, resized it to bellow 91.25%. (i'm afraid the lateral letters become too small)

> The other non-compulsory things to do are:
> 
> * The beginning of 03_iet-A is now a bit off. This sentence
> 
> ```
> This is the first part of a tutorial describing the Iteration/Expression Tree (IET), an intermediate representation based on Abstract Syntax Trees (AST) used in the Devito compiler. In this part we will show how to access and navigate the IET rooted in an Operator.
> ```

ok, fixed

> 
> Given the previous (existing and non-existing) notebooks, we should rewrite it simply as
> 
> ```
> In this tutorial we will show how to access and navigate the Iteration/Expression Tree (IET) rooted in an `Operator`.

fixed 

> ```
> 
> * This sentence
> 
> ```
> The reader of this tutorial is expected to be familiar with the basic Devito API (i.e., Grid, Function/TimeFunction, Operator, ...). Otherwise, the CFD tutorials in examples/cfd are a better place to start.
> ```
> 
> should/could be moved to 00_index (and replace "of this tutorial" with "of this series of tutorials")
> 

ok, fixed

> * Similarly, the beginning of 04_iet-B should be updated. From
> 
> ```
> This is the second part of a tutorial describing the Iteration/Expression Tree (IET), an intermediate representation based on Abstract Syntax Trees (AST) used in the Devito compiler. In this part we will learn how to build, compose, and transform IETs.
> ```
> 
> to something simpler
> 
> ```
> In this tutorial we will learn how to build, compose, and transform Iteration/Expression Trees (IETs).
> ```

roger that. fixed

FabioLuporini commented:
This is good to go now apart from the figure that really is unreadable. I created a new high-res PNG for you, which I'm attaching. Can you try with this one? 
![operator-diagram](https://user-images.githubusercontent.com/478143/57969349-6a6ad980-7976-11e9-89ae-8866e383e2e9.png)

FabioLuporini commented:
Can you also rebase on latest master please?

btrb commented:
> Can you also rebase on latest master please?

hm.. this is the first time i shall do that. i will create a `tutorial_usp-3` branch just in case, then proceed to rebase

FabioLuporini commented:
It's perfect now. Many thanks @btrb ! I really like it. Merged.
